---
title: >
  Putative Driver filtering
author:
- name: Krutika Gaonkar
  affiliation: 
  - &id1 Center for Data-Driven Discovery in Biomedicine, Children’s Hospital of Philadelphia, USA
  email: gaonkark@chop.email.edu
- name: Federico Marini
  affiliation: 
  - &id2 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
date: "`r Sys.Date()`"
package: "`r paste('annoFuse', packageVersion('annoFuse'))"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
vignette: >
  %\VignetteIndexEntry{Putative driver filtering}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{annoFuse}
  %\VignetteKeywords{Software, Transcription, Visualization, GeneExpression, Annotation,
    GeneFusionDetection, GUI, ReportWriting}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

# Introduction

In this vignette we demonstrate using annoFuse, to filter putative oncogene fusion from a filtered fusion list and also plot recurrent fusion, recurrently fused genes, and a summary plot across a large cancer cohort. We are using OpenPBTA data,which has 1028 pediatric RNA-seq samples spanning 19 different broad histologies.

# Step by step analysis

We start by loading `annoFuse` and the other required packages in the chunk below

```{r}
library("annoFuse")
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("qdapRegex"))
```

# Overview of the package
Here, we present annoFuse, an R package developed to annotate and filter expressed gene fusions, along with highlighting artifact filtered novel fusions.

 * Filtered fusion list `FilteredFusionAnnoFuse.tsv` was obtained from v16 OpenPBTA Star-Fusion, arriba and rsem expression (fpkm) files. 
 We used [FusionAnnotator](https://github.com/FusionAnnotator/FusionAnnotator/tree/v0.2.0) to annotate arriba files to specifically identify “red flag” fusions found in healthy tissues or in gene homology databases saved as column "annots" to match annotation in STAR-Fusion calls. 
 The merged dataset was processed using the following functions `fusion_standardization` -> `fusion_filtering_QC` -> `expression_filter_fusion` -> `annotate_fusion_calls` ). 
 Please refer to the filtering criterias available per function below:

![](Figure 1.png)

# Standard fusion call format requirements

  - Mandatory fields : "Sample","LeftBreakpoint", "RightBreakpoint", "FusionName","Gene1A", "Gene1B"
  
  For OpenPBTA samples we will also require some project specific fields to visualize and explore the data
  
  - Additional fields required for OpenPBTA project specific filter : "Kids_First_Biospecimen_ID","BreakpointLocation","broad_histology"


```{r}
fusion_calls <- read_tsv(system.file("extdata", "FilteredFusionAnnoFuse.tsv", package = "annoFuse"))

# distance are being removed here to capture all intergenic fusions in count
# distance within () were making them count as unique instead of the same fusion
fusion_calls$FusionName <- unlist(lapply(
  fusion_calls$FusionName, 
  function(x) rm_between(x, "(", ")", extract = FALSE)
))

cols_fusioncalls <- c("LeftBreakpoint", "RightBreakpoint", "FusionName",
                      "Gene1A", "Gene1B","Sample")

head(fusion_calls[,cols_fusioncalls])

```

These high quality fusion calls are considered known and putative oncogenic driver fusions if the fusion is previously reported in TCGA or fusions containing gene partners that are known oncogenes, tumor suppressor genes, COSMIC genes, or transcription factors. 


```{r}

# gene list
geneListReferenceDataTab <- read.delim(
  system.file("extdata", "genelistreference.txt", package = "annoFuse"),
  stringsAsFactors = FALSE
)

# fusion list
fusionReferenceDataTab <- read.delim(
  system.file("extdata", "fusionreference.txt", package = "annoFuse"),
  stringsAsFactors = FALSE
)

# filter for driver fusions
putative_driver_fusions <- fusion_driver(
  standardFusioncalls = fusion_calls, 
  annotated = TRUE, 
  geneListReferenceDataTab = geneListReferenceDataTab, 
  fusionReferenceDataTab = fusionReferenceDataTab
)
```

Putative Driver Fusions that are found in more than 4 histologies were removed from further processing as these fusions are considered artifact or commonly occurring from manual review.

```{r}
# checking if fusions are called in multiple histologies
# this will suggest these fusion calls are artifact or
# commonly occuring
found_in_morethan_4 <- groupcount_fusion_calls(
  putative_driver_fusions, 
  group = "broad_histology", 
  numGroup = 4
) %>% 
  arrange(desc(group.ct))

found_in_morethan_4
```

Other non-oncogenic (not annotated as oncogene/transcription factor/ TSG /kinase) fusions that are recurrent fusions called by both callers and unique to specific histology in this filtered dataset were of interest as well. 

However, non-oncogenic genes that are fused more than 5 times per sample were removed since we consider these are artifacts within our dataset from manual review of these fusions. 


```{r}
# To scavenge back non-oncogenic recurrent/unique per broad histology fusions
fusion_calls <- aggregate_fusion_calls(
  fusion_calls,
  removeother = TRUE,
  filterAnnots = "LOCAL_REARRANGEMENT|LOCAL_INVERSION"
)

# Keep
# 1. Called by at least n callers
fusion_calls.summary <- called_by_n_callers(fusion_calls, 
                                            numCaller = 2)

# OR
# 2. Found in at least n samples in each group
sample.count <- samplecount_fusion_calls(fusion_calls, 
                                         numSample = 2, 
                                         group = "broad_histology")

# Remove
# 1. non-oncogenic fusions that are in > numGroup
group.count <- groupcount_fusion_calls(fusion_calls, 
                                       group = "broad_histology", 
                                       numGroup = 1)

# 2. non-oncogenic multi-fused genes
fusion_recurrent5_per_sample <- fusion_multifused(fusion_calls, 
                                                  limitMultiFused = 5)

# filter fusion_calls to keep recurrent fusions from above sample.count and fusion_calls.summary
QCGeneFiltered_recFusion <- fusion_calls %>%
  dplyr::filter(FusionName %in% unique(c(sample.count$FusionName, fusion_calls.summary$FusionName)))

# filter QCGeneFiltered_recFusion to remove fusions found in more than 1 group and multifused gene per samples
QCGeneFiltered_recFusionUniq <- QCGeneFiltered_recFusion %>%
  dplyr::filter(!FusionName %in% group.count$FusionName) %>%
  dplyr::filter(!Gene1A %in% fusion_recurrent5_per_sample$GeneSymbol |
    !Gene2A %in% fusion_recurrent5_per_sample$GeneSymbol |
    !Gene1B %in% fusion_recurrent5_per_sample$GeneSymbol |
    !Gene2B %in% fusion_recurrent5_per_sample$GeneSymbol)
```

We were interested in visualizing fusion distributions that are within 2 different genes which is annotated as "Genic" in our standard fusion format.
Fusions that are in-frame technically should translate into proteins so we will also filter for only in-frame fusions for visualization

```{r}
putative_driver_fusions <- putative_driver_fusions %>%
  dplyr::bind_rows(QCGeneFiltered_recFusionUniq[, colnames(putative_driver_fusions)]) %>%
  # filter fusions found in more than 4 histologies
  dplyr::filter(
    !FusionName %in% found_in_morethan_4$FusionName,
    # remove intergenic and intragenic fusion
    BreakpointLocation == "Genic",
    Fusion_Type == "in-frame"
  ) %>%
  as.data.frame()


head(putative_driver_fusions)
```

The summary of fusions called can provide an overview of the genomic rearrangements within cohort in-terms of intra or inter chromosomal changes and biotypes of genes fused. We also provide distribution of kinase domain retained in 5` and 3` genes as well as overall distribution of annotation in each group. Here we've used broad_histology to graoup our cohorts. 

```{r fig.width=15, fig.height=15}
# PLot summary
plot_summary(standardFusioncalls = putative_driver_fusions, 
             groupby = "broad_histology")
```

Recurrent fusions and genes fused provide insigths into subtypes of samples within a cohort. Here we choose broad_histology as the grouping variable and plot the `n` participants (Kids_First_Participant_ID) with these fusions. 

```{r fig.width=6, fig.height=6}
# recurrently fused genes
plot_recurrent_genes(standardFusioncalls = putative_driver_fusions, 
                     groupby = "broad_histology", 
                     countID = "Kids_First_Participant_ID", 
                     plotn = 20)
```

```{r fig.width=6, fig.height=8}
# recurrent fusions
plot_recurrent_fusions(standardFusioncalls = putative_driver_fusions, 
                       groupby = "broad_histology", 
                       countID = "Kids_First_Participant_ID", 
                       plotn = 20)
```

# Session info {-}

```{r}
sessionInfo()
```

