---
title: >
  Putative Oncogenic Driver Fusion Filtering
author:
- name: Krutika Gaonkar
  affiliation: 
  - &id1 Center for Data-Driven Discovery in Biomedicine, Childrenâ€™s Hospital of Philadelphia, USA
  email: gaonkark@chop.email.edu
- name: Federico Marini
  affiliation: 
  - &id2 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz, Germany
  - Center for Thrombosis and Hemostasis (CTH), Mainz, Germany
  email: marinif@uni-mainz.de
date: "`r Sys.Date()`"
package: "`r paste('annoFuse', packageVersion('annoFuse'))"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
vignette: >
  %\VignetteIndexEntry{Putative driver filtering}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{annoFuse}
  %\VignetteKeywords{Software, Transcription, Visualization, GeneExpression, Annotation,
    GeneFusionDetection, GUI, ReportWriting}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

# Introduction

In this vignette, we demonstrate how to use annoFuse to filter putative oncogenic fusions from a filtered fusion list.
The filtered fusion list `FilteredFusionAnnoFuse.tsv` was obtained from the v16 release of [OpenPBTA](https://github.com/AlexsLemonade/OpenPBTA-analysis).
The steps are: `fusion_standardization` -> `fusion_filtering_QC` -> `annotate_fusion_calls`, followed by plotting recurrent fusions, recurrently fused genes, as well as a summary plot.


# Step by step analysis

We start by loading `annoFuse` and the other required packages in the chunk below

```{r}
library("annoFuse")
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("qdapRegex"))
```


```{r}
fusion_calls <- read_tsv(system.file("extdata", "FilteredFusionAnnoFuse.tsv", package = "annoFuse"))

# distance are being removed here to capture all intergenic fusions in count
# distance within () were making them count as unique instead of the same fusion
fusion_calls$FusionName <- unlist(lapply(
  fusion_calls$FusionName, 
  function(x) rm_between(x, "(", ")", extract = FALSE)
))

print("Raw calls from STARfusion and Arriba for PBTA")
table(fusion_calls$Caller)
```

If the fusion has been previously reported in TCGA or if fusions containing gene partners are known oncogenes, tumor suppressor genes, COSMIC genes, and/or transcription factors, then we consider the fusion call to be known oncogenic or putative oncogenic. 


```{r}
# aggreate caller to unique fusion calls per Sample
fusion_calls <- aggregate_fusion_calls(fusion_calls, removeother = FALSE)

# Add reference gene list containing known oncogenes, tumor suppressors, kinases, and transcription factors
geneListReferenceDataTab <- read.delim(
  system.file("extdata", "genelistreference.txt", package = "annoFuse"),
  stringsAsFactors = FALSE
)

# Add fusion list containing previously reported oncogenic fusions.
fusionReferenceDataTab <- read.delim(
  system.file("extdata", "fusionreference.txt", package = "annoFuse"),
  stringsAsFactors = FALSE
)

# filter for driver fusions
putative_driver_fusions <- fusion_driver(
  standardFusioncalls = fusion_calls, 
  annotated = TRUE, 
  geneListReferenceDataTab = geneListReferenceDataTab, 
  fusionReferenceDataTab = fusionReferenceDataTab
)
```

Putative Driver Fusions found in more than four distinct histologies were filtered out, as these fusions were considered likely artifactual.

```{r}
# checking if fusions are called in multiple histologies
# this will suggest these fusion calls are artifactual or
# commonly occuring
found_in_morethan_4 <- groupcount_fusion_calls(
  putative_driver_fusions, 
  group = "broad_histology", 
  numGroup = 4
) %>% 
  arrange(desc(group.ct))

found_in_morethan_4
```

Other non-oncogenic (not annotated as oncogene/transcription factor/tumor suppressor gene/kinase) fusions that are recurrent fusions, called by both callers, and are unique to a specific histology in this filtered dataset are of interest as well. 

However, non-oncogenic genes that are fused more than five times per sample were removed, as they were considered artifactual within our dataset from manual review. 


```{r}
# To scavenge back non-oncogenic recurrent/unique per broad histology fusions
scavenge_recurrent_non_onco <- fusion_calls <- aggregate_fusion_calls(
  fusion_calls,
  removeother = TRUE,
  filterAnnots = "LOCAL_REARRANGEMENT|LOCAL_INVERSION"
)

# Keep
# 1. Called by at least n callers
fusion_calls.summary <- called_by_n_callers(fusion_calls, 
                                            numCaller = 2)

# OR
# 2. Found in at least n samples in each group
sample.count <- samplecount_fusion_calls(fusion_calls, 
                                         numSample = 2, 
                                         group = "broad_histology")

# Remove
# 1. non-oncogenic fusions that are in > numGroup
group.count <- groupcount_fusion_calls(fusion_calls, 
                                       group = "broad_histology", 
                                       numGroup = 1)

# 2. non-oncogenic multi-fused genes
fusion_recurrent5_per_sample <- fusion_multifused(fusion_calls, 
                                                  limitMultiFused = 5)

# filter fusion_calls to keep recurrent fusions from above sample.count and fusion_calls.summary
QCGeneFiltered_recFusion <- fusion_calls %>%
  dplyr::filter(FusionName %in% unique(c(sample.count$FusionName, fusion_calls.summary$FusionName)))

# filter QCGeneFiltered_recFusion to remove fusions found in more than 1 group and multifused gene per samples
QCGeneFiltered_recFusionUniq <- QCGeneFiltered_recFusion %>%
  dplyr::filter(!FusionName %in% group.count$FusionName) %>%
  dplyr::filter(!Gene1A %in% fusion_recurrent5_per_sample$GeneSymbol |
    !Gene2A %in% fusion_recurrent5_per_sample$GeneSymbol |
    !Gene1B %in% fusion_recurrent5_per_sample$GeneSymbol |
    !Gene2B %in% fusion_recurrent5_per_sample$GeneSymbol)
```

Here, we visualize the distribution of fusions within two different genes that are annotated as "Genic" in our standard fusion format.
Fusions that are in-frame are predicted to translate into protein, so we also filter for only in-frame fusions for visualization.

```{r}
putative_driver_fusions <- putative_driver_fusions %>%
  dplyr::bind_rows(QCGeneFiltered_recFusionUniq[, colnames(putative_driver_fusions)]) %>%
  # filter fusions found in more than 4 histologies
  dplyr::filter(
    !FusionName %in% found_in_morethan_4$FusionName,
    # remove intergenic and intragenic fusion
    BreakpointLocation == "Genic",
    Fusion_Type == "in-frame"
  ) %>%
  as.data.frame()


head(putative_driver_fusions)
```

Recurrent fusions and genes fused provide insights into subtypes of samples within a cohort. 
Here we choose `broad_histology` as the grouping variable and plot the `n` participants (Kids_First_Participant_ID) with these fusions. 

```{r fig.width=6, fig.height=5}
# recurrently fused genes
plot_recurrent_genes(standardFusioncalls = putative_driver_fusions, 
                     groupby = "broad_histology", 
                     countID = "Kids_First_Participant_ID", 
                     plotn = 20)
```

```{r fig.width=6, fig.height=5}
# recurrent fusions
plot_recurrent_fusions(standardFusioncalls = putative_driver_fusions, 
                       groupby = "broad_histology", 
                       countID = "Kids_First_Participant_ID", 
                       plotn = 20)
```

The summary of fusions called can provide an overview of the genomic rearrangements within a cohort in-terms of intra- or inter-chromosomal changes and biotypes of genes fused. 
We also provide a distribution of whether the kinase domain was retained in 5` or 3` genes, as well as an overall distribution of annotation in each group. 
Here we've used `broad_histology` to group our cohorts. 

```{r fig.width=15, fig.height=15}
# PLot summary
plot_summary(standardFusioncalls = putative_driver_fusions, 
             groupby = "broad_histology")
```

# Session info {-}

```{r}
sessionInfo()
```
