---
title: "Single sample fusion annotation and filtering"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Project specific filtering}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```


# Introduction
Gene fusion events are a significant source of somatic variation across adult and pediatric cancers and have provided some of the most effective clinically relevant therapeutic targets, yet computational algorithms for fusion detection from RNA sequencing data show low overlap of predictions across methods. annoFuse provides a standardized filtering and annotating method for gene fusion calls from STAR-Fusion and Arriba by merging, filtering, and prioritizing putative oncogenic fusions across large cancer datasets, as demonstrated in the manuscript with the OpenPBTA project dataset.


### Generating StarFusion and arriba files
  * [STAR-Fusion] (https://github.com/STAR-Fusion/STAR-Fusion/wiki)
  * [Arriba] (https://arriba.readthedocs.io/en/latest/)

### Fusion format requirements

  * [STAR-Fusion star-fusion.fusion_predictions.tsv ](https://github.com/STAR-Fusion/STAR-Fusion/wiki#output-from-star-fusion)
  * [Arriba fusions.tsv](https://arriba.readthedocs.io/en/latest/output-files/)

### Rsem gene format
  * RSEM genes.results.gz

# Overview of the package
Here, we present annoFuse, an R package developed to annotate and filter expressed gene fusions, along with highlighting artifact filtered novel fusions. 

- Pre-processing: 
 * We used [FusionAnnotator](https://github.com/FusionAnnotator/FusionAnnotator/tree/v0.2.0) to annotate arriba files to specifically identify “red flag” fusions found in healthy tissues or in gene homology databases saved as column "annots" to match annotation in STAR-Fusion calls. Users can provide any annotation they seem useful for their dataset to this field and filter using the param `artifactFilter` in `annoFuse_single_sample()`

In this vignette we demonstrates using annoFuse, to filter out fusions known to be artifactual or with low read support and retain high-quality fusion calls from STAR-Fusion and Arriba for a single sample. We also annotate these high quality fusion calls and capture known and putative oncogenic driver fusions previously reported in TCGA or fusions containing gene partners that are known oncogenes, tumor suppressor genes, COSMIC genes, or transcription factors by running `annoFuse_single_sample()`

Additionally, in the vignette we annotate filtered fusion with pfam domains per Gene1A (Gene 5') and Gene1B (Gene 3') by running `get_Pfam_domain()` and visualize the breakpoint in BRAF causing the KIAA1549--BRAF fusion in this sample by running `plot_breakpoints()`


## Single sample Star Fusion and arriba standardization, annotation and filtering.

```{r}
suppressPackageStartupMessages(library("annoFuse"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("reshape2"))

# Run annoFuse for Single sample with default expression filter and FusionAnnotator red flag artifact filter
standardFusioncalls <- annoFuse::annoFuse_single_sample(
  # Example files are provided in extdata, at-least 1 fusionfile is required along with it's rsem expression file
  fusionfileArriba = system.file("extdata", "arriba_example.tsv", package = "annoFuse"),
  fusionfileStarFusion = system.file("extdata", "starfusion_example.tsv", package = "annoFuse"),
  expressionFile = system.file("extdata", "example.rsem.genes.results.gz", package = "annoFuse"),
  tumorID = "BS_W97QQYKQ",
  # multiple read flag values for filtering using FusionAnnotator values
  artifactFilter = "GTEx_Recurrent|DGD_PARALOGS|Normal|BodyMap|ConjoinG",
  # keep all in-frame , frameshift and other types of Fusion_Type
  readingFrameFilter = "in-frame|frameshift|other",
  # keep all fusions with atleast 1 junction read support
  junctionReadCountFilter = 1,
  # keep only fusions where spanningFragCount-junctionReadCountFilter less than equal to 10
  spanningFragCountFilter = 10,
  # keep read throughs
  readthroughFilter = FALSE
)

# Add domain level information for fusion
# read in pfamDataBioMart with pfam and gene boundaries from UCSC pfam and biomaRt package
bioMartDataPfam <- readRDS(system.file("extdata", "pfamDataBioMart.RDS", package = "annoFuse"))
annDomain <- annoFuse::get_Pfam_domain(
  standardFusioncalls = standardFusioncalls,
  bioMartDataPfam = bioMartDataPfam,
  # partial overlapping domains are retained == "Partial" with keepPartialAnno=TRUE; if keepPartialAnno=FALSE then domain retained status == "No"
  keepPartialAnno = TRUE
)

# plot BRAF breakpoint in sample for KIAA1549--BRAF fusion
kiaabraf <- annDomain$Gene1B[which(annDomain$Gene1B$FusionName == "KIAA1549--BRAF" & annDomain$Gene1B$Gene1B == "BRAF"), ] %>% dplyr::filter(!is.na(DESC))

# read in exonsToPlot with exon and gene boundaries from gencode.v27.primary_assembly.annotation.gtf.gz
exons <- readRDS(system.file("extdata", "exonsToPlot.RDS", package = "annoFuse"))
```

## Plot breakpoint

```{r , fig.width=10, fig.height=5}
plot_breakpoints(domainDataFrame = kiaabraf, exons = exons, geneposition = "Right") + theme_publication(base_size = 12)
```



